manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,740 @@
+that are created using the \stdmoduleref{Atomic} module.
+
+\subsection{s:happens_before}{Happens-before relation}
+
+Let us imagine that the OCaml programs are executed by an abstract machine that
+executes one action at a time, arbitrarily picking one of the available domains
+at each step. We classify actions into two: \emph{inter-domain} and
+\emph{intra-domain}. An inter-domain action is one which can be observed and be
+influenced by actions on other domains. There are several inter-domain actions:
+
+\begin{itemize}
+  \item Reads and writes of atomic and non-atomic locations.
+  \item Spawn and join of domains.
+  \item Operations on mutexes.
+\end{itemize}
+
+\emph{(TODO: Include semaphores and condition variables in inter-domain
+actions.)}
+
+On the other hand, an intra-domain actions can neither be observed nor
+influence the execution of other domains. Examples include evaluating an
+arithmetic expression, calling a function, etc. The memory model specification
+ignores such intra-domain actions. In the sequel, we use the term action to
+indicate inter-domain actions.
+
+A totally ordered list of actions executed by the abstract machine is called
+the \emph{execution trace}. There might be several execution traces for a given
+program due to non-determinism.
+
+For a given execution trace, we define an irreflexive, transitive
+\emph{happens-before relation} that captures the causality between actions in
+the OCaml program. The happens-before relation is defined as follows:
+
+\begin{itemize}
+  \item We define the order in which a domain executes its actions as the
+    \emph{program order}. If an action "x" precedes another action "y" in
+    program order, then "x" precedes "y" in happens-before order.
+  \item If "x" is an initialising write to a memory location (atomic or
+    otherwise) and "y" is any read or write to that memory location, then "x"
+    precedes "y" in happens-before order.
+  \item If "x" is a write to an atomic location and "y" is a subsequent read or
+    write to that memory location in the execution trace, then "x" precedes "y"
+    in happens-before order. For atomic locations, "compare_and_set",
+    "fetch_and_add", "exchange", "incr" and "decr" are considered to perform
+    both a read and a write.
+  \item If "x" is "Domain.spawn f" and "y" is the first action in the newly
+    spawned domain executing "f", then "x" precedes "y" in happens-before
+    order.
+  \item If "x" is the last action in a domain "d" and "y" is "Domain.join
+    d", then "x" precedes "y" in happens-before order.
+  \item If "x" is an unlock operation on a mutex, and "y" is any subsequent
+    operation on the mutex in the execution trace, then "x" precedes "y" in
+    happens-before order.
+\end{itemize}
+
+\subsection{s:datarace}{Data race}
+
+In a given trace, two actions are said to be \emph{conflicting} if they access
+the same non-atomic location, and at least one is a write.
```

Comment:
Conflicting accesses originate from different concurrent agents (threads?)
```suggestion
the same non-atomic location, are performed by two different threads, and at least one is a write.
```

Reply:
Not necessarily. Two accesses from the same thread may also be conflicting. But such a pair will not have a data race since the program order between the pair of accesses implies happens-before order. 

The definition of conflicting access closely follows the definition (9) from the PLDI paper on the OCaml memory model: https://kcsrk.info/papers/pldi18-memory.pdf. I've tried not to be creative here.